### PR TITLE
build: apply publishing config after project evaluation

### LINKED ISSUE #3620: Broken dependency resolution in 0.13.0-SNAPSHOT and Spark 3.2
Hi all!

I'm trying to use iceberg's snapshot release (to make use of the better spark 3.2 support), but I'm getting the following error when running spark-shell (full log [here](https://gist.github.com/racevedoo/3530bce04364629f81217e78dea66071))
```
spark-shell --repositories https://repository.apache.org/content/repositories/snapshots \
 --packages org.apache.iceberg:iceberg-spark-3.2-runtime:0.13.0-SNAPSHOT,org.postgresql:postgresql:42.2.10,software.amazon.awssdk:url-connection-client:2.17.68,software.amazon.awssdk:sts:2.17.68,software.amazon.awssdk:s3:2.17.68,software.amazon.awssdk:dynamodb:2.17.68,software.amazon.awssdk:glue:2.17.68


:: problems summary ::
:::: WARNINGS
                module not found: org.slf4j#slf4j-api;working@mymachine

        ==== local-m2-cache: tried

          file:/home/rafael.acevedo/.m2/repository/org/slf4j/slf4j-api/working@mymachine/slf4j-api-working@mymachine.pom

          -- artifact org.slf4j#slf4j-api;working@mymachine!slf4j-api.jar:

          file:/home/rafael.acevedo/.m2/repository/org/slf4j/slf4j-api/working@mymachine/slf4j-api-working@mymachine.jar

        ==== local-ivy-cache: tried

          /home/rafael.acevedo/.ivy2/local/org.slf4j/slf4j-api/working@mymachine/ivys/ivy.xml

          -- artifact org.slf4j#slf4j-api;working@mymachine!slf4j-api.jar:

          /home/rafael.acevedo/.ivy2/local/org.slf4j/slf4j-api/working@mymachine/jars/slf4j-api.jar

        ==== central: tried

          https://repo1.maven.org/maven2/org/slf4j/slf4j-api/working@mymachine/slf4j-api-working@mymachine.pom

          -- artifact org.slf4j#slf4j-api;working@mymachine!slf4j-api.jar:

          https://repo1.maven.org/maven2/org/slf4j/slf4j-api/working@mymachine/slf4j-api-working@mymachine.jar

        ==== spark-packages: tried

          https://repos.spark-packages.org/org/slf4j/slf4j-api/working@mymachine/slf4j-api-working@mymachine.pom

          -- artifact org.slf4j#slf4j-api;working@mymachine!slf4j-api.jar:

          https://repos.spark-packages.org/org/slf4j/slf4j-api/working@mymachine/slf4j-api-working@mymachine.jar

        ==== repo-1: tried

          https://repository.apache.org/content/repositories/snapshots/org/slf4j/slf4j-api/working@mymachine/slf4j-api-working@mymachine.pom

          -- artifact org.slf4j#slf4j-api;working@mymachine!slf4j-api.jar:

          https://repository.apache.org/content/repositories/snapshots/org/slf4j/slf4j-api/working@mymachine/slf4j-api-working@mymachine.jar

                ::::::::::::::::::::::::::::::::::::::::::::::

                ::          UNRESOLVED DEPENDENCIES         ::

                ::::::::::::::::::::::::::::::::::::::::::::::

                :: org.slf4j#slf4j-api;working@mymachine: not found

                ::::::::::::::::::::::::::::::::::::::::::::::


:::: ERRORS
        unknown resolver null

        unknown resolver null

        unknown resolver public

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver sbt-chain

        unknown resolver null

        unknown resolver sbt-chain

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null

        unknown resolver null


:: USE VERBOSE OR DEBUG MESSAGE LEVEL FOR MORE DETAILS
Exception in thread "main" java.lang.RuntimeException: [unresolved dependency: org.slf4j#slf4j-api;working@mymachine: not found]
        at org.apache.spark.deploy.SparkSubmitUtils$.resolveMavenCoordinates(SparkSubmit.scala:1447)
        at org.apache.spark.util.DependencyUtils$.resolveMavenDependencies(DependencyUtils.scala:185)
        at org.apache.spark.deploy.SparkSubmit.prepareSubmitEnvironment(SparkSubmit.scala:308)
        at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:898)
        at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
        at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
        at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
        at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1043)
        at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1052)
        at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```

Running with iceberg 0.12.1 works fine:
```
spark-shell --repositories https://repository.apache.org/content/repositories/snapshots --packages org.apache.iceberg:iceberg-spark3-runtime:0.12.1,org.postgresql:postgresql:42.2.10,software.amazon.awssdk:url-connection-client:2.17.68,software.amazon.awssdk:sts:2.17.68,software.amazon.awssdk:s3:2.17.68,software.amazon.awssdk:dynamodb:2.17.68,software.amazon.awssdk:glue:2.17.68
```

From my investigation, slf4j is now a runtime dependency (should be included with spark) since #3335 (https://github.com/apache/iceberg/blob/master/spark/v3.2/build.gradle#L181). In my case, dependency resolution is failing for some reason

### PR DESCRIPTION
**Pull request by racevedoo**
Since we refactored the spark projects to have separate folders/projects
per spark version (in 811af43), gradle was evaluating the maven publish
plugin config (in `deploy.gradle`) before it applied the `shadowJar`
configs required for spark runtime projects, causing the generated POM
to include dependencies when it shouldn't (as we're publishing a shadow
jar).

This fixes it by delaying the maven publishing config to
`afterEvaluate`, when gradle has already applied the shadowJar plugin
for projects.

Closes #3620

### CODE REVIEW

**Comment by racevedoo** [comment]
@rdblue I found the root cause of the issue we discussed in #3642 (the pom generation issue). I believe this is the proper fix

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Awesome. Thank you @racevedoo! I had looked at this a little bit, but had no idea what the cause was.

**Comment by racevedoo** [comment]
> Awesome. Thank you @racevedoo! I had looked at this a little bit, but had no idea what the cause was.

No worries! I had to dig pretty deep to find this hehehe. Anything I must do before this gets merged?

**Comment by kbendick** [comment]
> No worries! I had to dig pretty deep to find this hehehe.

Thank you so much for this deep dive @racevedoo! It's greatly appreciated.

> Anything I must do before this gets merged?

I don't think so. If it's been approved by a PMC member (like @/rdblue) or even a committer, they're most likely just giving it more time for people to take a look for themselves and give any feedback required before some committer comes and merges it.

So the only thing you'd need to do would be respond to questions or concerns that people bring up in the mean time 😄

File: deploy.gradle
Comment (kbendick) [inline deploy.gradle:54]: Small nit / learning opportunity: There is a trailing white space after the `->`. I think that's what's making this diff look different than the other ones (that show that they are leading space changes only).

Generally speaking, the smaller the diff change the easier it is for people to cherry-pick and backport into their own forks. It's probably not as big of a deal for `deploy.gradle`, but I thought I'd take this opportunity to let you know as you get more involved with the project. 😄 

Usually, the goal is to simply keep the number of changes as small as possible so that if people `git cherry-pick`, they're less likely to have issues with the diff applying correctly. This file doesn't merit that level of scrutiny imho (though I currently don't maintain a fork).

That said, there's probably no major reason to change it. In some of the existing stuff in the file, there's a terminal space afterwards and some there is not. So this has just been for knowledge sake (if you weren't already aware). 😄

**Review by kbendick — APPROVED** [review]
+1 LGTM. Thanks again for going on this deep dive.

File: deploy.gradle
Comment (jackye1995) [inline deploy.gradle:129]: nit: extra space

**Review by jackye1995 — APPROVED** [review]
thank you, really appreciate the fix!!

**Comment by jackye1995** [comment]
I think everyone approves, this is super critical contribution, thank you! There were just some nitpicking, @racevedoo up to you if you would like to clean those up. I will merge after some time.

File: deploy.gradle
Comment (rdblue) [reply deploy.gradle:54]: I don't think that this was added by this PR. I think everything in the block was indented with no other change (at least, from ignore whitespace diff).

**Comment by rdblue** [comment]
I was waiting for CI checks to pass. I also validated this manually and the pom files look correct again. I think this is ready to merge when Jack is satisfied.

**Comment by rdblue** [comment]
I realized I can just remove the extra newline through the github UI. I'll merge this. Thanks!

**Pull request merged**
